src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -169,86 +189,139 @@ const std::vector<std::string>& AbstractLQPNode::output_column_names() const {
   return left_child()->output_column_names();
 }
 
-const std::vector<ColumnID>& AbstractLQPNode::output_column_ids_to_input_column_ids() const {
+const std::vector<LQPColumnReference>& AbstractLQPNode::output_column_references() const {
   /**
-   * This function has to be overwritten if columns or their order are in any way redefined by this Node.
-   * Examples include Projections, Aggregates, and Joins.
+   * Default implementation of output_column_references() will return the ColumnReferences of the left_child if it exists,
+   * otherwise will pretend that all Columns originate in this node.
+   * Nodes with both children need to override this as the default implementation can't cover their behaviour.
    */
-  DebugAssert(left_child() && !right_child(),
-              "Node has no or two inputs and therefore needs to override this function.");
 
-  if (!_output_column_ids_to_input_column_ids) {
-    _output_column_ids_to_input_column_ids.emplace(output_column_count());
-    std::iota(_output_column_ids_to_input_column_ids->begin(), _output_column_ids_to_input_column_ids->end(),
-              ColumnID{0});
+  if (!_output_column_references) {
+    Assert(!left_child() || !right_child(), "Nodes that have two children must override this method");
```

Comment:
If `right_child() != nullptr`, we have a big problem and this assertion makes things worse. `Assert(!right_child(), ...)`?

Reply:
agreed